File: python/src/abslang/formatters/table.py
"""Output formatters: table (rich), JSON, JUnit XML."""

import json
import xml.sax.saxutils as saxutils

from rich.console import Console
from rich.table import Table
from rich.text import Text

from ..runner import RunResult


def _escape_xml(s: str) -> str:
    return saxutils.escape(s)


def format_table(result: RunResult) -> str:
    """Format a single RunResult as a rich table."""
    console = Console(force_terminal=True, width=100)
    table = Table(title="ABS — Results", show_header=True, header_style="bold")
    table.add_column("#", width=4)
    table.add_column("Step", width=36)
    table.add_column("Result", width=10)
    table.add_column("", width=9)

    for step in result.steps:
        desc = f"{step.behavior.actor} {step.behavior.action}"
        if step.behavior.target:
            desc += f" {step.behavior.target}"
        content = step.behavior.content
        if isinstance(content, str):
            desc += f' "{content[:20]}"'
        elif content:
            desc += f" {json.dumps(content)[:20]}"

        if step.sent:
            table.add_row(str(step.step), desc[:34], "→", "sent")
        elif step.matched:
            table.add_row(str(step.step), desc[:34], "✅", "match")
            for ev in step.evaluations:
                ev_desc = ev.type[:28]
                ev_result = "✅" if ev.passed else "❌"
                ev_label = "pass" if ev.passed else "FAIL"
                table.add_row("", f"  └─ {ev_desc}", ev_result, ev_label)
        else:
            table.add_row(str(step.step), desc[:34], "❌", "no match")

    for ev in result.chain_evaluations:
        ev_desc = f"chain: {ev.type}"[:34]
        ev_result = "✅" if ev.passed else "❌"
        ev_label = "pass" if ev.passed else "FAIL"
        table.add_row("C", ev_desc, ev_result, ev_label)

    # Capture table as string
    with console.capture() as capture:
        console.print(table)
        console.print(f"\nSession: {result.session}")
        console.print(f"Agent: {result.agent}")
        status = "✅ PASSED" if result.passed else "❌ FAILED"
        style = "green" if result.passed else "red"
        console.print(f"Result: [{style}]{status}[/{style}]")
        console.print(
            f"Steps: {result.steps_matched}/{result.steps_total} matched · "
            f"{result.evaluations_passed}/{result.evaluations_total} evaluations passed"
        )

        if not result.passed:
            console.print()
            console.print("[red]❌ Some evaluations failed:[/red]")
            for step in result.steps:
                for ev in step.evaluations:
                    if not ev.passed:
                        console.print(f"[red]  Step {step.step} — {ev.type}: {ev.reason}[/red]")
            for ev in result.chain_evaluations:
                if not ev.passed:
                    console.print(f"[red]  Chain — {ev.type}: {ev.reason}[/red]")

    return capture.get()


def format_json_output(result: RunResult) -> str:
    """Format a RunResult as JSON."""
    return json.dumps({
        "session": result.session,
        "agent": result.agent,
        "passed": result.passed,
        "steps_total": result.steps_total,
        "steps_matched": result.steps_matched,
        "evaluations_total": result.evaluations_total,
        "evaluations_passed": result.evaluations_passed,
        "trace": [
            {
                "step": s.step,
                "behavior": {
                    "actor": s.behavior.actor,
                    "action": s.behavior.action,
                    "target": s.behavior.target,
                    "content": s.behavior.content,
                },
                "matched": s.matched,
                "sent": s.sent,
                "observed": {
                    "actor": s.observed.actor,
                    "action": s.observed.action,
                    "target": s.observed.target,
                    "content": s.observed.content,
                } if s.observed else None,
                "evaluations": [
                    {
                        "type": e.type,
                        "passed": e.passed,
                        "score": e.score,
                        "reason": e.reason,
                    }
                    for e in s.evaluations
                ],
            }
            for s in result.steps
        ],
        "chain_evaluations": [
            {"type": e.type, "passed": e.passed, "score": e.score, "reason": e.reason}
            for e in result.chain_evaluations
        ],
    }, indent=2, default=str)


def format_junit(result: RunResult) -> str:
    """Format a RunResult as JUnit XML."""
    all_evals = []
    for s in result.steps:
        for e in s.evaluations:
            all_evals.append((s.step, e, False))
    for e in result.chain_evaluations:
        all_evals.append((None, e, True))

    failures = sum(1 for _, e, _ in all_evals if not e.passed)
    total = len(all_evals)

    xml = '<?xml version="1.0" encoding="UTF-8"?>\n'
    xml += f'<testsuite name="ABS: {_escape_xml(result.session)}" tests="{total}" failures="{failures}" errors="0">\n'

    for step_num, ev, is_chain in all_evals:
        name = f'{"Chain" if is_chain else f"Step {step_num}"}: {ev.type}'
        xml += f'  <testcase classname="ABS" name="{_escape_xml(name)}" time="0">\n'
        if not ev.passed:
            xml += f'    <failure message="{_escape_xml(ev.reason)}">\n'
            xml += f'      Type: {_escape_xml(ev.type)}\n'
            xml += f'      Reason: {_escape_xml(ev.reason)}\n'
            xml += f'    </failure>\n'
        xml += f'  </testcase>\n'

    xml += '</testsuite>\n'
    return xml

File: typescript/src/formatters/table.ts
import { RunResult } from "../runner";
import chalk from "chalk";

export function formatTable(
  result: RunResult,
  rowsTotal?: number,
  rowsPassed?: number,
  rowVars?: Record<string, any>
): string {
  const lines: string[] = [];

  lines.push("┌──────────────────────────────────────────────────────────────┐");
  lines.push("│  ABS — Results                                               │");
  lines.push("├──────────────────────────────────────────────────────────────┤");
  lines.push(`│  Session:  ${result.session.padEnd(52)}│`);
  lines.push(`│  Agent:    ${result.agent.padEnd(52)}│`);

  if (rowsTotal !== undefined) {
    lines.push(
      `│  Dataset:  ${String(rowsTotal).padEnd(52)} rows│`
    );
  }

  const status = result.passed ? chalk.green("✅ PASSED") : chalk.red("❌ FAILED");
  lines.push(`│  Result:   ${status.padEnd(62)}│`);

  if (rowsTotal !== undefined && rowsPassed !== undefined) {
    lines.push(
      `│  Rows:     ${String(rowsPassed)}/${String(rowsTotal)} passed · ${String(rowsTotal - rowsPassed)} failed`.padEnd(64) + "│"
    );
  }

  lines.push(
    `│  Steps:    ${result.stepsMatched}/${result.stepsTotal} matched · ${result.evaluationsPassed}/${result.evaluationsTotal} evaluations passed`.padEnd(64) + "│"
  );

  lines.push("├────┬────────────────────────────────────┬──────────┬─────────┤");
  lines.push("│  # │ Step                               │ Result   │         │");
  lines.push("├────┼────────────────────────────────────┼──────────┼─────────┤");

  for (const step of result.steps) {
    const num = String(step.step).padStart(2);
    const desc = stepDescription(step).substring(0, 34).padEnd(34);

    if (step.sent) {
      lines.push(`│ ${num} │ ${desc} │    →     │   sent  │`);
    } else if (step.matched) {
      lines.push(`│ ${num} │ ${desc} │    ✅    │  match  │`);
    } else {
      lines.push(`│ ${num} │ ${desc} │    ❌    │  no match│`);
    }

    for (const ev of step.evaluations) {
      const evDesc = ev.type.substring(0, 28).padEnd(28);
      const evStatus = ev.inconclusive
        ? "   ⚠️    │  incon  │"
        : ev.passed ? "   ✅    │   pass  │" : "   ❌    │  FAIL  │";
      lines.push(`│    │   └─ ${evDesc} │ ${evStatus}`);
    }
  }

  // Chain evaluations
  for (const ev of result.chainEvaluations) {
    const desc = `chain: ${ev.type}`.substring(0, 34).padEnd(34);
    const status = ev.inconclusive
      ? "   ⚠️    │  incon  │"
      : ev.passed ? "   ✅    │   pass  │" : "   ❌    │  FAIL  │";
    lines.push(`│  C │ ${desc} │ ${status}`);
  }

  lines.push("└────┴────────────────────────────────────┴──────────┴─────────┘");

  if (!result.passed) {
    lines.push("");
    lines.push(chalk.red("❌ Some evaluations failed:"));
    for (const step of result.steps) {
      for (const ev of step.evaluations) {
        if (!ev.passed && !ev.inconclusive) {
          lines.push(chalk.red(`  Step ${step.step} — ${ev.type}: ${ev.reason}`));
        }
      }
    }
    for (const ev of result.chainEvaluations) {
      if (!ev.passed && !ev.inconclusive) {
        lines.push(chalk.red(`  Chain — ${ev.type}: ${ev.reason}`));
      }
    }
  }

  const inconclusive = [
    ...result.steps.flatMap((s) => s.evaluations),
    ...result.chainEvaluations,
  ].filter((e) => e.inconclusive).length;
  if (inconclusive > 0) {
    lines.push("");
    lines.push(chalk.yellow(`⚠️  ${inconclusive} evaluation(s) marked inconclusive (downstream of a blocking failure).`));
  }

  return lines.join("\n");
}

function stepDescription(step: any): string {
  const b = step.behavior;
  const content =
    typeof b.content === "string"
      ? b.content
      : b.content
        ? JSON.stringify(b.content)
        : "";
  return `${b.actor} ${b.action}${b.target ? " " + b.target : ""} "${content.substring(0, 20)}"`;
}

export function formatJson(result: RunResult): string {
  return JSON.stringify(
    {
      session: result.session,
      agent: result.agent,
      passed: result.passed,
      steps_total: result.stepsTotal,
      steps_matched: result.stepsMatched,
      evaluations_total: result.evaluationsTotal,
      evaluations_passed: result.evaluationsPassed,
      trace: result.steps.map((s) => ({
        step: s.step,
        behavior: {
          actor: s.behavior.actor,
          action: s.behavior.action,
          target: s.behavior.target,
          content: s.behavior.content,
        },
        matched: s.matched,
        sent: s.sent,
        observed: s.observed,
        evaluations: s.evaluations,
      })),
      chain_evaluations: result.chainEvaluations,
    },
    null,
    2
  );
}

export function formatJunit(result: RunResult): string {
  const allEvals = [
    ...result.steps.flatMap((s) =>
      s.evaluations.map((e) => ({ ...e, step: s.step, isChain: false }))
    ),
    ...result.chainEvaluations.map((e) => ({ ...e, isChain: true })),
  ];

  const failures = allEvals.filter((e) => !e.passed).length;
  const total = allEvals.length;

  let xml = `<?xml version="1.0" encoding="UTF-8"?>\n`;
  xml += `<testsuite name="ABS: ${escapeXml(result.session)}" tests="${total}" failures="${failures}" errors="0">\n`;

  let caseNum = 0;
  for (const step of result.steps) {
    for (const ev of step.evaluations) {
      caseNum++;
      xml += `  <testcase classname="ABS" name="Step ${step.step}: ${escapeXml(ev.type)}" time="0">\n`;
      if (ev.inconclusive) {
        xml += `    <skipped message="${escapeXml(ev.reason)}" />\n`;
      } else if (!ev.passed) {
        xml += `    <failure message="${escapeXml(ev.reason)}">\n`;
        xml += `      Type: ${escapeXml(ev.type)}\n`;
        xml += `      Reason: ${escapeXml(ev.reason)}\n`;
        xml += `    </failure>\n`;
      }
      xml += `  </testcase>\n`;
    }
  }

  for (const ev of result.chainEvaluations) {
    caseNum++;
    xml += `  <testcase classname="ABS" name="Chain: ${escapeXml(ev.type)}" time="0">\n`;
    if (ev.inconclusive) {
      xml += `    <skipped message="${escapeXml(ev.reason)}" />\n`;
    } else if (!ev.passed) {
      xml += `    <failure message="${escapeXml(ev.reason)}">\n`;
      xml += `      Type: ${escapeXml(ev.type)}\n`;
      xml += `      Reason: ${escapeXml(ev.reason)}\n`;
      xml += `    </failure>\n`;
    }
    xml += `  </testcase>\n`;
  }

  xml += `</testsuite>\n`;
  return xml;
}

function escapeXml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}
